### ISSUE TITLE
OOMs on the Jest docusaurus site

### ISSUE THREAD

**Issue by orta** [shikijs/twoslash#35]
Maybe a leak in Shiki?

![Screen Shot 2021-06-17 at 8 27 26 PM](https://user-images.githubusercontent.com/49038/122460803-7d639400-cfaa-11eb-98f2-66105178ddc6.png)

**Comment by orta** [comment]
Somewhere in the stack, there must be a `str += char` or something, all the new objects are concatenated strings:

![Screen Shot 2021-06-17 at 9 08 31 PM](https://user-images.githubusercontent.com/49038/122465640-3ed0d800-cfb0-11eb-90c4-ae5a65b882cc.png)

E.g. this string is 3000 links which I think are just 3000 `+=`s of one char:

![Screen Shot 2021-06-17 at 9 10 48 PM](https://user-images.githubusercontent.com/49038/122465913-8c4d4500-cfb0-11eb-8e6e-587bded95240.png)

Which is this tmlanguage  in shiki:

![Screen Shot 2021-06-17 at 9 12 12 PM](https://user-images.githubusercontent.com/49038/122466120-cd455980-cfb0-11eb-91ef-b9cdcc66c194.png)

**Comment by orta** [comment]
I think it's more likely that the issue lays in the remark-shiki-twoslash somehow re-creating the highlighters, but I can't prove it.

**Comment by frencojobs** [comment]
Came across this bug today. This is a severe bug since this makes twoslash completely unusable. I think a lot more people might be using docusaurus with jest and this breaks the setup entirely. 

The error is also scarier because I don't understand any of the logs and what's happening.

**Comment by frencojobs** [comment]
Could this be related to this https://github.com/facebook/docusaurus/issues/4978?

**Comment by orta** [comment]
Yeah,  I thought we were correctly caching the highlighters with https://github.com/shikijs/twoslash/blob/main/packages/remark-shiki-twoslash/src/index.ts#L61-L62

But maybe caching the highlighter doesn't cache the syntax JSONs 

It's not a problem I've had on Gatsby with the TS website which has many hundreds of twoslash code samples, so there's a chance it's a docusaurus level but I think we can still find a way at this library level

**Comment by orta** [comment]
I think #43 might have fixed this, 
![Screen Shot 2021-06-26 at 4 51 32 PM](https://user-images.githubusercontent.com/49038/123518659-d44e2500-d69e-11eb-8444-825a27adc672.png)

**Comment by frencojobs** [comment]
I just checked it. Indeed, it is fixed. Thanks!

**Comment by frencojobs** [comment]
Probably not really the fault of twoslash. But https://github.com/remotion-dev/remotion/commit/2f00e61cbe5728362c888fb5053266b4e45c4755 is still experiencing this. I'm guessing those garbage collectible weakmaps can only help a bit, and the problem is more about https://github.com/facebook/docusaurus/pull/4997.

**Comment by frencojobs** [comment]
This is reproducible with `docusaurus build` while `docusaurus start` yields no problem during development whatsoever.

**Comment by frencojobs** [comment]
Managed to reproduce the behavior with 80 (it was fine till 50) markdown files of this content on a plain docusaurus template.
```md
```ts twoslash
type CreateMutable<Type> = {
  -readonly [Property in keyof Type]: Type[Property];
};

type LockedAccount = {
  readonly id: string;
  readonly name: string;
};

type UnlockedAccount = CreateMutable<LockedAccount>;
\```
```

The content is probably not important, just put it out there for reference.

**Comment by frencojobs** [comment]
When I increased the memory to `4096MB`, the error is changed to something else. 

```log
Error: EMFILE: too many open files, open 'D:\a\packages\docs\node_modules\shiki\languages\xml.tmLanguage.json'
Error: EMFILE: too many open files, open 'D:\a\packages\docs\node_modules\shiki\languages\xml.tmLanguage.json'
Error: EMFILE: too many open files, open 'D:\a\packages\docs\node_modules\shiki\languages\tsx.tmLanguage.json'
Error: EMFILE: too many open files, open 'D:\a\packages\docs\node_modules\shiki\languages\xml.tmLanguage.json'
Error: EMFILE: too many open files, open 'D:\a\packages\docs\node_modules\shiki\languages\xml.tmLanguage.json'
Error: EMFILE: too many open files, open 'D:\a\packages\docs\node_modules\shiki\languages\xml.tmLanguage.json'
Error: EMFILE: too many open files, open 'D:\a\packages\docs\node_modules\shiki\languages\groovy.tmLanguage.json'
Error: EMFILE: too many open files, open 'D:\a\packages\docs\node_modules\shiki\languages\xml.tmLanguage.json'
Error: EMFILE: too many open files, open 'D:\a\packages\docs\node_modules\shiki\languages\xml.tmLanguage.json'
Error: EMFILE: too many open files, open 'D:\a\packages\docs\node_modules\shiki\languages\xml.tmLanguage.json'
Error: EMFILE: too many open files, open 'D:\a\packages\docs\node_modules\shiki\languages\erlang.tmLanguage.json'
```

Maybe XML is related to tsx, but I don't have these groovy and erlang languages used in the code. Does shiki read all these files necessary or not??

Ref: https://github.com/remotion-dev/remotion/runs/2931935721

**Comment by orta** [comment]
Hrm, still seems a tad off - I would have thought that caching the shiki highlighter would mean those files only get set up once, but if there's dupes in there - it obviously isn't

**Comment by frencojobs** [comment]
I'm really confused right now. Does the remark-shiki-twoslash code contain a logic bug or something? I set up a counter server and got this (on my 80 files docusaurus setup with remark plugin).

```json5
{
  "getHighlighter": 182, // how many times shiki.getHighlighter is called
  "setCache": 0, // how many times the data is set to weakmap
  "getFromCache": 0, // how many times the highlighters are taken from weakmap
  "fileHit": 1 // how many times the file is called
}
```

**Comment by frencojobs** [comment]
I found out that the caches are set but every remark API call try to load the highlighter and set the cache also. It seems like the remark's async handlers are being run parallel and the getHighlighter is taking quite some time so caching becomes useless as all of the functions are trying to load highlighters into cache at the same time.

When I add a `console.log` line on the start and end(before return) of the `highlightersFromSettings` function, I got all the `before` lines at once instead of getting them in pairs.

```js
before 0 // <- length of highlightCache map
before 0
before 0
// ...
after 1 // <- length of highlightCache map
after 2 // <- why duplicated when there's only one setting
after 3
```

There's also the problem with duplicated keys in the map.

The correct behavior would be,

```js
before 0
after 1 // only hit once because the function shouldn't reach to end if the cache exists
before 1 
before 1
before 1
```
